firmware/library/L1_Drivers/ST7066U.hpp

```diff
@@ -0,0 +1,196 @@
+
+// The file contains the driver class for the ST7066U Parallel LCD Driver.
+//
+// Usage:
+//      Initialize the driver by calling the constructor:
+//       ST7066U(BusMode busMode, DisplayLines numLines,
+//               Font fontStyle, ControlPins_t *pins)
+//      If the desired busMode is 4-bits, then initialize D3, D2, D1,
+//      and D0 in ControlPins_t as NULL.
+//      Characters can be generated on the display using the function:
+//          void displayText(const char *text, CursorPosition_t pos)
+#pragma once
+
+#include "L0_LowLevel/LPC40xx.h"
+#include "L1_Drivers/gpio.hpp"
+
+class ST7066UInterface
+{
+ public:
+    struct CursorPosition_t;
+
+    virtual void init()                                               = 0;
+    virtual void clearDisplay()                                       = 0;
+    virtual void toggleDisplay(bool on)                               = 0;
+    virtual void toggleCursor(bool on)                                = 0;
+    virtual void setCursorDirection(bool increment)                   = 0;
+    virtual void displayText(const char * text, CursorPosition_t pos) = 0;
+
+ protected:
+    // Data transfer operation types.
+    enum class WriteOperationType : uint8_t
+    {
+        kCommandWrite = 0,
+        kDataWrite    = 1
+    };
+
+    enum class Command : uint8_t
+    {
+        // Clears all characters on the display.
+        kClearDisplay = 0x01,
+        // Resets the cursor  to line 0, position 0.
+        kResetCursor = 0x02,
+        // Sets the cursor direction to write leftwards.
+        kCursorDirectionLeft = 0x04,
+        // Sets the cursor direction to write rightwards.
+        kCursorDirectionRight = 0x06,
+        // Toggles the display OFF.
+        kToggleDisplayOff = 0x08,
+        // Toggles the display ON, with cursor blink OFF.
+        kToggleDisplayOn = 0x0C,
+        // Toggles the diplay ON, with cursor blink ON.
+        kToggleCursorOn = 0x0F,
+        // Default mask for display configuration.
+        kDefaultConfig = 0x20,
+        // Base cursor address position for line 0.
+        kLineAddress0 = 0x80,
+        // Base cursor address position for line 1.
+        kLineAddress1 = 0xC0
+    };
+
+    virtual inline void writeNibble(WriteOperationType operation,
+                                    uint8_t nibble)             = 0;
+    virtual inline void writeByte(WriteOperationType operation,
+                                  uint8_t byte)                 = 0;
+    virtual inline void writeCommand(Command cmd)               = 0;
+    virtual inline void writeCommand(uint8_t cmd)               = 0;
+    virtual inline void writeData(uint8_t data)                 = 0;
+    virtual inline void setCursorPosition(CursorPosition_t pos) = 0;
+    virtual inline void resetCursorPosition()                   = 0;
+};
+
+class ST7066U : public ST7066UInterface
```

Comment:
This commit should not include implementation details such as this. Save for the next commit.